straxen/plugins/event_processing.py

```diff
@@ -733,58 +734,137 @@ def cs2_to_e(self, x):
 
 
 @export
+@strax.takes_config(
+    strax.Option(name='pre_s1_area_threshold', default=1e3,
+                 help='Only take S1s larger than this into account '
+                      'when calculating PeakShadow [PE]'),
+    strax.Option(name='pre_s2_area_threshold', default=1e4,
+                 help='Only take S2s larger than this into account '
+                      'when calculating PeakShadow [PE]'),
+    strax.Option(name='deltatime_exponent', default=-1.0,
+                 help='The exponent of delta t when calculating shadow'),
+    strax.Option(name='time_window_backward', default=int(1e9),
+                 help='Search for peaks casting shadow in this time window [ns]')
+)
 class EventShadow(strax.Plugin):
     """
-    This plugin can calculate shadow at event level.
-    It depends on peak-level shadow.
-    The event-level shadow is its first S2 peak's shadow.
-    If no S2 peaks, the event shadow will be nan.
-    It also gives the position infomation of the previous S2s
-    and main peaks' shadow.
+    This plugin can calculate shadow for main S1 and main S2 in events.
+    It also gives the position information of the previous peaks.
+    References:
+        * v0.1.2 reference: xenon:xenonnt:ac:prediction:shadow_ambience
     """
-    __version__ = '0.0.8'
+    __version__ = '0.1.2'
     depends_on = ('event_basics', 'peak_basics', 'peak_shadow')
     provides = 'event_shadow'
+    data_kind = 'events'
     save_when = strax.SaveWhen.EXPLICIT
 
     def infer_dtype(self):
-        dtype = [('s1_shadow', np.float32, 'main s1 shadow [PE/ns]'),
-                 ('s2_shadow', np.float32, 'main s2 shadow [PE/ns]'),
-                 ('shadow', np.float32, 'shadow of event [PE/ns]'),
-                 ('pre_s2_area', np.float32, 'previous s2 area [PE]'),
-                 ('shadow_dt', np.int64, 'time difference to the previous s2 [ns]'),
-                 ('shadow_index', np.int32, 'max shadow peak index in event'),
-                 ('pre_s2_x', np.float32, 'x of previous s2 peak causing shadow [cm]'),
-                 ('pre_s2_y', np.float32, 'y of previous s2 peak causing shadow [cm]'),
-                 ('shadow_distance', np.float32, 'distance to the s2 peak with max shadow [cm]')]
+        dtype = []
+        for main_peak, main_peak_desc in zip(['s1_', 's2_'], ['main S1', 'main S2']):
+            # previous S1 can only cast time shadow, previous S2 can cast both time & position shadow
+            for key in ['s1_time', 's2_time', 's2_position']:
+                type_str = key.split('_')[0]
+                tp_desc = key.split('_')[1]
+                dtype.append(((f'largest {tp_desc} shadow casting from previous {type_str} to {main_peak_desc} [PE/ns]', f'{main_peak}shadow_{key}'), np.float32))
```

Comment:
still I have a question about` nearest previous large s1`.

- what do you mean by large? If area_size is above pre_s1_area_threshold , then such s1s are called as large s1s?
- what do you mean by nearest? in which time_window? maybe time_window_backward(1sec)? Maybe nearest mean minimum time_difference w.r.t main S1, right?

Reply:
- The S1 larger than `pre_s1_area_threshold` is referred to as large S1 in codes. 
- Yes!